src/persist/src/operators/source.rs

```diff
@@ -0,0 +1,184 @@
+
+//! A Timely Dataflow operator that mirrors a persisted stream.
+
+use std::sync::mpsc::{self, TryRecvError};
+
+use timely::dataflow::operators::generic::operator;
+use timely::dataflow::operators::{Concat, ToStream};
+use timely::dataflow::{Scope, Stream};
+use timely::Data;
+
+use crate::indexed::runtime::{StreamReadHandle, StreamWriteHandle};
+use crate::indexed::TailEvent;
+use crate::{operators, Token};
+
+/// A Timely Dataflow operator that mirrors a persisted stream.
+pub trait PersistedSource<G: Scope<Timestamp = u64>, K: Data, V: Data> {
+    /// Emits a snapshot of the persisted stream taken as of this call and tails
+    /// any new data added to the persisted stream after that.
+    fn persisted_source(
+        &mut self,
+        read: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    );
+}
+
+impl<G, K, V> PersistedSource<G, K, V> for G
+where
+    G: Scope<Timestamp = u64>,
+    K: Data + Send,
+    V: Data + Send,
+{
+    fn persisted_source(
+        &mut self,
+        token: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    ) {
+        let (_, read) = token.into_inner();
+        let (tail_tx, tail_rx) = mpsc::channel();
+        let tail_fn = Box::new(move |e| {
+            // TODO: If send fails, it means the operator is no longer running.
+            // We should probably allow the tail to deregister itself.
+            let _ = tail_tx.send(e);
+        });
+
+        let err_new = match read.tail(tail_fn) {
+            Ok(_) => operator::empty(self),
+            Err(err) => vec![(err.to_string(), 0, 1)].to_stream(self),
+        };
+
+        // TODO: Plumb the name of the stream down through the handles and use
+        // it for the operator.
+        let ok_new = operator::source(self, "PersistedSource", |capability, info| {
+            // TODO: Hook the activator up to the callback instead of
+            // TryRecvError::Empty.
+            let activator = self.activator_for(&info.address[..]);
+            let mut cap = Some(capability);
+            move |output| {
+                let mut done = false;
+                if let Some(cap) = cap.as_mut() {
+                    let mut session = output.session(cap);
+                    match tail_rx.try_recv() {
+                        Ok(e) => match e {
+                            TailEvent::Records(mut records) => {
+                                session.give_vec(&mut records);
+                                activator.activate();
+                            }
+                            TailEvent::Sealed(ts) => {
+                                cap.downgrade(&ts);
+                            }
+                        },
+                        Err(TryRecvError::Empty) => {
+                            activator.activate();
```

Comment:
maybe we want to activate_after a few seconds here?

Reply:
Done.